platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryFailureCollector.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.ivyservice.ivyresolve;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Data structure to collect repository failures during dependency resolution.
+ *
+ * @see DynamicVersionResolver
+ * @see RepositoryChainComponentMetaDataResolver
+ */
+class RepositoryFailureCollector {
+    private final Collection<Throwable> failures = new ArrayList<>();
+    private boolean fatalError;
+
+    void addFailure(Throwable failure) {
+        failures.add(failure);
+    }
+
+    Collection<Throwable> getFailures() {
+        return failures;
+    }
+
+    boolean hasFatalError() {
+        return fatalError;
+    }
+
+    void toggleFatalError() {
```

Comment:
toggle implies that calling this twice would return the state back to the original state. `markFatalError` or `markAsFatal` would be a better name.